src/dmd/aliasthis.d

```diff
@@ -63,50 +63,61 @@ extern (C++) final class AliasThis : Dsymbol
 
 Expression resolveAliasThis(Scope* sc, Expression e, bool gag = false)
 {
-    AggregateDeclaration ad = isAggregate(e.type);
-    if (ad && ad.aliasthis)
+    for (AggregateDeclaration ad = isAggregate(e.type); ad;)
     {
-        uint olderrors = gag ? global.startGagging() : 0;
-        Loc loc = e.loc;
-        Type tthis = (e.op == TOK.type ? e.type : null);
-        e = new DotIdExp(loc, e, ad.aliasthis.ident);
-        e = e.expressionSemantic(sc);
-        if (tthis && ad.aliasthis.needThis())
+        if (ad.aliasthis)
         {
-            if (e.op == TOK.variable)
+            uint olderrors = gag ? global.startGagging() : 0;
+            Loc loc = e.loc;
+            Type tthis = (e.op == TOK.type ? e.type : null);
+            e = new DotIdExp(loc, e, ad.aliasthis.ident);
+            e = e.expressionSemantic(sc);
+            if (tthis && ad.aliasthis.needThis())
             {
-                if (auto fd = (cast(VarExp)e).var.isFuncDeclaration())
+                if (e.op == TOK.variable)
                 {
-                    // https://issues.dlang.org/show_bug.cgi?id=13009
-                    // Support better match for the overloaded alias this.
-                    bool hasOverloads;
-                    if (auto f = fd.overloadModMatch(loc, tthis, hasOverloads))
+                    if (auto fd = (cast(VarExp)e).var.isFuncDeclaration())
                     {
-                        if (!hasOverloads)
-                            fd = f;     // use exact match
-                        e = new VarExp(loc, fd, hasOverloads);
-                        e.type = f.type;
-                        e = new CallExp(loc, e);
-                        goto L1;
+                        // https://issues.dlang.org/show_bug.cgi?id=13009
+                        // Support better match for the overloaded alias this.
+                        bool hasOverloads;
+                        if (auto f = fd.overloadModMatch(loc, tthis, hasOverloads))
+                        {
+                            if (!hasOverloads)
+                                fd = f;     // use exact match
+                            e = new VarExp(loc, fd, hasOverloads);
+                            e.type = f.type;
+                            e = new CallExp(loc, e);
+                            goto L1;
+                        }
                     }
                 }
+                /* non-@property function is not called inside typeof(),
+                 * so resolve it ahead.
+                 */
+                {
+                    int save = sc.intypeof;
+                    sc.intypeof = 1; // bypass "need this" error check
+                    e = resolveProperties(sc, e);
+                    sc.intypeof = save;
+                }
+            L1:
+                e = new TypeExp(loc, new TypeTypeof(loc, e));
```

Comment:
Bummer we still have a `goto`. I think this can be resolved by putting lines 105 and 106 in a `scope(exit)` at the top of the scope, and using a `bool` that conditionally executes lines 95-103.

Reply:
The general philosophy in dmd is to not mix refactorings with patches. This chunk of code is not touched by this PR (even though github presents it as such).